spanner/metrics.go

```diff
@@ -0,0 +1,474 @@
+	// These attributes need to be added to only few of the metrics
+	metricsDetails = map[string]metricInfo{
+		metricNameOperationCount: {
+			additionalAttrs: []string{
+				metricLabelKeyStatus,
+			},
+			recordedPerAttempt: false,
+		},
+		metricNameOperationLatencies: {
+			additionalAttrs: []string{
+				metricLabelKeyStatus,
+			},
+			recordedPerAttempt: false,
+		},
+		metricNameAttemptLatencies: {
+			additionalAttrs: []string{
+				metricLabelKeyStatus,
+			},
+			recordedPerAttempt: true,
+		},
+		metricNameAttemptCount: {
+			additionalAttrs: []string{
+				metricLabelKeyStatus,
+			},
+			recordedPerAttempt: true,
+		},
+	}
+
+	// Generates unique client ID in the format go-<random UUID>@<hostname>
+	generateClientUID = func() (string, error) {
+		hostname := "localhost"
+		hostname, err := os.Hostname()
+		if err != nil {
+			return "", err
+		}
+		return uuid.NewString() + "@" + strconv.FormatInt(int64(os.Getpid()), 10) + "@" + hostname, nil
+	}
+
+	exporterOpts = []option.ClientOption{}
+)
+
+type metricInfo struct {
+	additionalAttrs    []string
+	recordedPerAttempt bool
+}
+
+type builtinMetricsTracerFactory struct {
+	enabled             bool
+	isDirectPathEnabled bool
+
+	// To be called on client close
+	shutdown func()
+
+	// attributes that are specific to a client instance and
+	// do not change across different function calls on client
+	clientAttributes []attribute.KeyValue
+
+	operationLatencies metric.Float64Histogram
+	attemptLatencies   metric.Float64Histogram
+	operationCount     metric.Int64Counter
+	attemptCount       metric.Int64Counter
+}
+
+func detectClientLocation(ctx context.Context) string {
+	resource, err := gcp.NewDetector().Detect(ctx)
+	if err != nil {
+		return "global"
+	}
+	for _, attr := range resource.Attributes() {
+		if attr.Key == semconv.CloudRegionKey {
+			return attr.Value.AsString()
+		}
+	}
+	// If region is not found, return global
+	return "global"
+}
+
+func newBuiltinMetricsTracerFactory(ctx context.Context, dbpath string, metricsProvider metric.MeterProvider) (*builtinMetricsTracerFactory, error) {
+	clientUID, err := generateClientUID()
+	if err != nil {
+		log.Printf("built-in metrics: generateClientUID failed: %v. Using empty string in the %v metric atteribute", err, metricLabelKeyClientUID)
+	}
+	project, instance, database, err := parseDatabaseName(dbpath)
+	if err != nil {
+		return nil, err
+	}
+
+	tracerFactory := &builtinMetricsTracerFactory{
+		enabled: false,
+		clientAttributes: []attribute.KeyValue{
+			attribute.String(monitoredResLabelKeyProject, project),
+			attribute.String(monitoredResLabelKeyInstance, instance),
+			attribute.String(metricLabelKeyDatabase, database),
+			attribute.String(metricLabelKeyClientUID, clientUID),
+			attribute.String(metricLabelKeyClientName, clientName),
+			// Skipping instance config until we have a way to get it
+			attribute.String(monitoredResLabelKeyInstanceConfig, "unknown"),
```

Comment:
What does this mean to customers? Do they see this in somewhere?

Reply:
Since getting the value for instance config involves admin RPC call and there is quota limit per project, we decided to keep the label but have unknown as value for now.
Java Reference: https://github.com/googleapis/java-spanner/pull/3125/files#diff-c896ca5e475a075f267b0294d92328dcd35db60bdd408600bf05354119650be1R85